File: 214-shortest-palindrome/214-shortest-palindrome.py
class Solution:
    def KMP(self, text: str) -> list:
        i, j, kmp = 0, -1, [0] * (len(text) + 1)
        kmp[0] = -1
        while i < len(text):
            while j != -1 and text[i] != text[j]:
                j = kmp[j]
            i, j = i + 1, j + 1
            kmp[i] = j
        return kmp
        
    def shortestPalindrome(self, s: str) -> str:
        temp = s[::-1]
        text = s + "#" + temp
        kmp = self.KMP(text)
        pos = kmp[len(text)]
        ans = s[pos:]
        return ans[::-1] + s

File: 214-shortest-palindrome/214-shortest-palindrome.java
class Solution {
    private int[] KMP(String text) {
        int[] kmp = new int[text.length() + 1];
        int i = 0, j = -1;
        kmp[0] = -1;
        while(i < text.length()) {
            while(j != -1 && text.charAt(i) != text.charAt(j))
                j = kmp[j];
            ++i;
            ++j;
            kmp[i] = j;
        }
        return kmp;
    }
    public String shortestPalindrome(String s) {
        String rev = new StringBuilder(s).reverse().toString();
        String text = s + "#" + rev;
        int n = text.length();
        int[] kmp = KMP(text);
        int pos = kmp[n];
        String ans = s.substring(pos);
        String temp = new StringBuilder(ans).reverse().toString();
        ans = temp;
        return ans + s;
    }
}